reference/decoder.go

```diff
@@ -0,0 +1,287 @@
+	return &decoder{
+		byteDecoderFactory: byteDecoderFactory,
+		legacyDecoder:      byteDecoderFactory.LegacyDecoder(),
+		defaultDecoder:     byteDecoderFactory.DefaultDecoder(),
+
+		options: options,
+	}
+}
+
+func (v decoder) Decode(ref string) (Global, error) {
+	schemaPos := strings.IndexRune(ref, ':')
+	if schemaPos >= 0 {
+		decoder, err := v.parseSchema(ref[:schemaPos], ref)
+		if err != nil {
+			return Global{}, err
+		}
+		return v.parseReference(ref[schemaPos+1:], decoder)
+	}
+
+	// try to parse the legacy format
+	if v.options&AllowLegacy != 0 && len(ref) >= 2*len(LegacyDomainName)+1 {
+		domainPos := strings.IndexRune(ref, '.')
+		if domainPos >= len(LegacyDomainName) && ref[domainPos+1:] == LegacyDomainName {
+			return v.parseLegacyAddress(ref, domainPos)
+		}
+	}
+
+	return v.parseReference(ref, v.defaultDecoder)
+}
+
+func (v decoder) parseLegacyAddress(ref string, domainPos int) (Global, error) {
+	var result Global
+
+	w := result.addressLocal.asWriter()
+	_, err := v.legacyDecoder(ref[:domainPos], w)
+
+	switch {
+	case err != nil:
+		break
+	case !w.isFull():
+		err = errors.New("insufficient length")
+	case result.addressLocal.getScope() != 0: // there is no scope for legacy
+		err = errors.New("invalid scope")
+	case !result.tryConvertToSelf():
+		err = errors.New("invalid self reference")
+	default:
+		return result, nil
+	}
+	return result, fmt.Errorf("unable to parse legacy reference, %s: ref=%s", err.Error(), ref)
+}
+
+func (v decoder) parseSchema(scheme, refFull string) (ByteDecodeFunc, error) {
+	parts := strings.Split(scheme, "+")
+	switch len(parts) {
+	case 1:
+		if parts[0] == SchemaV1 {
+			return v.defaultDecoder, nil
+		}
+	case 2:
+		switch {
+		case parts[0] == SchemaV1:
+			parts[0] = parts[1]
+		case parts[1] == SchemaV1:
+			//
+		default:
+			return nil, fmt.Errorf("unsupported scheme: scheme=%s, ref=%s", scheme, refFull)
+		}
+	default:
+		return nil, fmt.Errorf("invalid scheme: scheme=%s", scheme)
+	}
+	decoder := v.byteDecoderFactory.GetByteDecoder(parts[0])
+	if decoder == nil {
+		return nil, fmt.Errorf("unknown decoder: scheme=%s, decoder=%s, ref=%s", scheme, parts[0], refFull)
+	}
+	return decoder, nil
+}
+
+func (v decoder) parseAuthority(ref string) (authority string, remaining string) {
+	if len(ref) < 3 || ref[:2] != "//" {
+		return "", ref
+	}
+	ref = ref[2:]
+
+	pos := strings.IndexRune(ref, '/')
+	if pos < 0 {
+		return ref, ""
+	}
+
+	return ref[:pos], ref[pos+1:]
+}
+
+func (v decoder) parseReference(refFull string, byteDecoder ByteDecodeFunc) (Global, error) {
+	_, ref := v.parseAuthority(refFull)
+	if len(ref) == 0 {
+		return Global{}, fmt.Errorf("empty reference body: ref=%s", refFull)
+	}
+
+	parityPos := strings.IndexRune(ref, '/')
+	var parity []byte
+	switch {
+	case parityPos == 0:
+		return Global{}, fmt.Errorf("empty reference body: ref=%s", refFull)
+	case parityPos > 0:
+		buf := bytes.NewBuffer(make([]byte, 0, pulseAndScopeSize))
```

Comment:
please apply the recent spec update - parity must be prefixed with "2"

Reply:
fixed